repositories/transfercheck_enrichment.go

```diff
@@ -49,44 +51,100 @@ func NewTransferCheckEnrichmentRepository(blobRepository BlobRepository, bucket
 	}
 }
 
-// Expects a CSV file with two columns: IP range and country code (ISO 3166-1 alpha-3) containing both ipv4 and ipv6 ranges.
-func (r *TransferCheckEnrichmentRepository) setupIpCountryRanges(ctx context.Context) error {
+// IP country methods
+
+func (r *TransferCheckEnrichmentRepository) GetIPCountry(ctx context.Context, ip netip.Addr) (string, error) {
 	tracer := utils.OpenTelemetryTracerFromContext(ctx)
 	ctx, span := tracer.Start(
 		ctx,
-		"repositories.TransferCheckEnrichmentRepository.setupIpCountryRanges",
+		"repositories.TransferCheckEnrichmentRepository.GetIPCountry",
 	)
 	defer span.End()
+
+	return r.findCountryDichotomy(ctx, ip)
+}
+
+func (r *TransferCheckEnrichmentRepository) findCountryDichotomy(ctx context.Context, ip netip.Addr) (string, error) {
+	ranges, err := r.getIpCountryRanges(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	left := 0
+	right := len(ranges) - 1
+
+	for right >= left {
+		mid := left + (right-left)/2
+
+		ipRange := ranges[mid].ipRange
+		if ipRange.Contains(ip) {
+			return ranges[mid].country, nil
+		} else if ip.Less(ipRange.Addr()) {
+			right = mid - 1
+		} else {
+			left = mid + 1
+		}
+	}
+
+	return "", nil
+}
+
+func (r *TransferCheckEnrichmentRepository) getIpCountryRanges(ctx context.Context) ([]ipCountryRange, error) {
 	r.muCountries.Lock()
 	defer r.muCountries.Unlock()
-	if time.Now().Before(r.countryRangesExpireAt) {
-		return nil
+
+	if time.Now().After(r.countryRangesExpireAt) {
+		ranges, err := r.readIpCountryRangesFromBlob(ctx)
+		if err != nil {
+			return nil, err
+		}
+		r.ipCountryRanges = ranges
+		r.countryRangesExpireAt = time.Now().Add(IP_RANGE_FILES_EXPIRATION)
 	}
 
+	return r.ipCountryRanges, nil
+}
+
+// Expects a CSV file with two columns: IP range and country code (ISO 3166-1 alpha-3) containing both ipv4 and ipv6 ranges.
+func (r *TransferCheckEnrichmentRepository) readIpCountryRangesFromBlob(ctx context.Context) ([]ipCountryRange, error) {
+	tracer := utils.OpenTelemetryTracerFromContext(ctx)
+	ctx, span := tracer.Start(
+		ctx,
+		"repositories.TransferCheckEnrichmentRepository.readIpCountryRangesFromBlob",
+	)
+	defer span.End()
+
 	file, err := r.blobRepository.GetBlob(ctx, r.bucket, IP_COUNTRY_RANGE_FILE)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer file.ReadCloser.Close()
 	fileReader := csv.NewReader(file.ReadCloser)
 	record, err := fileReader.Read()
+
+	_, span = tracer.Start(
```

Comment:
This is where I have a big blank spot in the traces I see that timeout, so add a span.